rust/pattern/constraint/thing_constraint.rs

```diff
@@ -0,0 +1,433 @@
+        HasConstraint {
+            type_: Some(UnboundVariable::hidden().type_(type_name).into_type()),
+            attribute: variable,
+        }
+    }
+
+    pub fn from_variable(variable: ThingVariable) -> Self {
+        HasConstraint {
+            type_: None,
+            attribute: variable,
+        }
+    }
+}
+
+impl Display for HasConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("has")?;
+        if let Some(type_) = &self.type_ {
+            write!(f, " {}", &type_.label.as_ref().unwrap().scoped_type)?;
+        }
+
+        use Reference::*;
+        match self.attribute.reference {
+            Name(_) => write!(f, " {}", self.attribute.reference),
+            Anonymous(_) => write!(f, " {}", self.attribute.value.as_ref().unwrap()),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ValueConstraint {
+    pub predicate: Predicate,
+    pub value: Value,
+}
+
+impl ValueConstraint {
+    pub fn into_constraint(self) -> Constraint {
+        self.into_thing_constraint().into_constraint()
+    }
+
+    pub fn into_thing_constraint(self) -> ThingConstraint {
+        ThingConstraint::Value(self)
+    }
+
+    pub fn new(predicate: Predicate, value: Value) -> ValueConstraint {
+        if predicate.is_substring() && !matches!(value, Value::String(_)) {
+            panic!("");
+        }
+        ValueConstraint { predicate, value }
+    }
+}
+
+impl Display for ValueConstraint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.predicate == Predicate::Eq && !matches!(self.value, Value::Variable(_)) {
+            write!(f, "{}", self.value)
+        } else {
+            write!(f, "{} {}", self.predicate, self.value)
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Predicate {
+    // equality
+    Eq,
+    Neq,
+    Gt,
+    Gte,
+    Lt,
+    Lte,
+    // substring
+    Contains,
+    Like,
+}
+
+impl Predicate {
+    pub fn is_equality(&self) -> bool {
+        use Predicate::*;
+        matches!(self, Eq | Neq | Gt | Gte | Lt | Lte)
+    }
+
+    pub fn is_substring(&self) -> bool {
+        use Predicate::*;
+        matches!(self, Contains | Like)
+    }
+}
+
+impl From<String> for Predicate {
+    fn from(string: String) -> Self {
+        use Predicate::*;
+        match string.as_str() {
+            "=" => Eq,
+            _ => todo!(),
+        }
+    }
+}
+
+impl Display for Predicate {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Predicate::*;
+        write!(
+            f,
+            "{}",
+            match self {
+                Eq => "=",
+                Neq => "!=",
+                _ => todo!(),
+            }
+        )
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Value {
+    Long(i64),
+    Double(f64),
+    Boolean(bool),
+    String(String),
+    DateTime(NaiveDateTime),
+    Variable(Box<ThingVariable>),
```

Comment:
Non-`Box` would make `ThingVariable` a recursive type: in the case like `$x = $y` the `ThingVariable` representing `$x` has a `ValueConstraint` with `Predicate::Eq` and value = `ThingVariable` representing `$y`.

Reply:
That type structure is impossible, as it has infinite size. `Option` is essentially a tagged union, and so is `Value`. 

`Value` has the size of at least `ThingVariable` + `bool` (`Option<ThingVariable>`);
`ValueConstraint` has the size of at least `Predicate` + `Value`, i.e. `ThingVariable` + a bunch.
`ThingVariable` has the size of all its contents, which means its size is greater than the size of `ValueConstraint`, which in turn is larger than a `ThingVariable`, which is absurd.